cql3/column_condition.cc

```diff
@@ -72,15 +72,12 @@ int is_satisfied_by(cql3::expr::oper_t op, const abstract_type& cell_type,
         }
 }
 
-// Read the list index from key and check that list index is not
-// negative. The negative range check repeats Cassandra behaviour.
-uint32_t read_and_check_list_index(const cql3::raw_value_view& key) {
+// Read the list index from key. Negative values are allowed, like
+// in PostgresQL (but unlike Cassandra).
+int32_t read_list_index(const cql3::raw_value_view& key) {
```

Comment:
I'm not thrilled about this change. I don't understand why PostgresQL is our gold standard when Cassandra should be.
But there is another reason why it's good for c[-1] to not be an error - it allows later to make the expression more general, like c[x] and we don't throw if x<0 (this is important while filtering, so we skip the item which has x=-1 instead of stopping the entire request.
By the way, in Alternator we actually have different error-handling if the invalid operand came from the user's request (in that case we throw) or from the database (in that case we just skip this data while filtering).

Reply:
> I'm not thrilled about this change. I don't understand why PostgresQL is our gold standard when Cassandra should be. 

PostgresQL is a representative of the SQL world, of which CQL is a dialect. CQL is widely inconsistent (just as in this case). Our goal should be to appeal to the wider audience of database users, not the narrower world of Cassandra users. Cassandra error handling is more or less random since it has many fewer users and the dialect is very constrained.

> But there is another reason why it's good for c[-1] to not be an error - it allows later to make the expression more general, like c[x] and we don't throw if x<0 (this is important while filtering, so we skip the item which has x=-1 instead of stopping the entire request. By the way, in Alternator we actually have different error-handling if the invalid operand came from the user's request (in that case we throw) or from the database (in that case we just skip this data while filtering).